amgcl/detail/inverse.hpp

```diff
@@ -31,44 +31,67 @@ THE SOFTWARE.
  * \brief  Inverse of a dense matrix.
  */
 
-#include <vector>
+#include <array>
 #include <algorithm>
 #include <cassert>
+#include <numeric>
+#include <utility>
 #include <amgcl/util.hpp>
 #include <amgcl/value_type/interface.hpp>
 
 namespace amgcl {
 namespace detail {
 
-    template <typename value_type>
-    static void inverse(int n, value_type *A, value_type *t) {
+    template <typename value_type, int n>
+    static void static_inverse(value_type *A, value_type *t) {
```

Comment:
`detail::inverse` is also used in parts of the code unrelated to `static_matrix`, for example in `deflated_solver` [here](https://github.com/ddemidov/amgcl/blob/c781719e2f961c161a7ff0cf2c168ac6949ac510/amgcl/deflated_solver.hpp#L162).

Reply:
I think if the allocation of the permutation vector is moved to the calling code, then p could be a `std::array` for inverting a static matrix, and `std::vector` otherwise.

Reply:
Hmm, the test suite compiles and runs on my machine. Looks like that code path is not tested.

I can either move creation of `p` to the caller code, or make a separate function where n is not known at compile time.

Reply:
I had to switch on `-DAMGCL_BUILD_EXAPMLES=TRUE` to see the problem.

Reply:
> Looks like that code path is not tested.

It is not tested indeed, sorry about that.

> I can either move creation of p to the caller code, or make a separate function where n is not known at compile time.

I think the former option would result in less code duplication.

Reply:
I have moved allocation to the calling code.